### ISSUE TITLE
The prowjob_state_transitions metric is imprecise and unavailable for non-kubernetes agent ProwJobs.

### ISSUE THREAD

**Issue by cjwagner** [kubernetes/test-infra#19459]
**What happened**:
Plank/Prow-controller-manager's `prowjob_state_transitions` metric doesn't seem to be properly reporting the state transitions per loop (or per 30 seconds in the case of the controller manager). See this graph that shows the results for the following query:
```
prowjob_state_transitions{job_name="ci-test-infra-bazel", state="success"}
```
![Screenshot from 2020-10-05 17-58-13](https://user-images.githubusercontent.com/5334145/95147232-01376200-0735-11eb-93dc-f19cf4275089.png)

Notice that the values incorrectly start at an unreasonably high value whenever reset (due to restarting thanks to a version bump).
There are also state transitions that are missed by the metrics, especially the `triggered` state due to how we detect transitions.

**What you expected to happen**:
I'd expect this metric to start at a more reasonably value (e.g. 1). Retroactively determining how many jobs transitioned state like we do here: https://github.com/kubernetes/test-infra/blob/89788a6d3d306540d65858deed2be5e8e0049fbf/prow/kube/metrics.go#L126-L138
causes the unreasonably high starting value and seems unnecessarily complicated compared to simply incrementing the counter when we actually transition each job.
Retroactively checking for state transitions is also potentially problematic for the prow-controller-manager since it does this every 30 seconds rather than once per sync loop like plank does. This makes it possible for transitions to be missed if a job transitions multiple times within 30 seconds.

/area prow/plank
/assign

**Comment by cjwagner** [comment]
It looks like the large starting values are not usually detected because the prometheus `rate()` function handles breaks in monotonicity as a single increment and the initial surge of increments happens very quickly, making it unlikely that the prometheus server scrapes during the initial surge (which would cause it to actually count the remaining increments in the surge as opposed to considering the entire surge an increment of 1).
This is a rare race so it isn't as impactful as missing transitions, but incrementing the counter when we actually transition each job should address both problems. That being said I still see two issues:
1. For jobs that are in the triggered state, how do we distinguish between the first time they are seen and any requeues while they are still in the triggered state?
2. We are only generating these ProwJob metrics for `agent: kubernetes` jobs. If we want to have metrics for all ProwJobs this logic should probably live in a separate component. If this lives in a separate component we'll need to keep using a cache of the previous states like we do now, but we could make the metrics more accurate by using a controller runtime manager rather than a 30 second polling loop. (1) would not be a problem if we did this.

@alvaroaleman I think we want a separate prowjob-metrics controller for this which is more involved than my original fix and probably isn't worth the investment right now, but would be more worthwhile once pipeline support is improved. WDYT?

**Comment by alvaroaleman** [comment]
> It looks like the large starting values are not usually detected because the prometheus rate() function handles breaks in monotonicity as a single increment and the initial surge of increments happens very quickly, making it unlikely that the prometheus server scrapes during the initial surge (which would cause it to actually count the remaining increments in the surge as opposed to considering the entire surge an increment of 1).

That would still be fine because then `rate()` would just handle two breaks of monotonicity instad of one? The bug here is IMHO that we consider "not seen" an increase but it happens to not be a problem.

re 1: Not sure I understand, jobs get created with `.status.state` set to `triggered` already, why would we count they multiple times?

re 2.: I think this was a design decision. The jenkins controller e.G. also publishes this metric for jenkins jobs. The idea was apparently that the `agent` is inferred from the `job` label.

> @alvaroaleman I think we want a separate prowjob-metrics controller for this which is more involved than my original fix and probably isn't worth the investment right now, but would be more worthwhile once pipeline support is improved. WDYT?

We already have such a component in prow/cmd/exporter. I don't really care a lot if we export this metric from each individual controller for its own agent or for all agents on the exporter

**Comment by cjwagner** [comment]
> That would still be fine because then `rate()` would just handle two breaks of monotonicity instad of one?

There is only one break in monotonicity (larger value to smaller value), the second scrape would see a significant change in the value, but it would be monotonic (smaller value to larger value) so the rate() function would count the full size of the value increase, not just a count of 1.


> The bug here is IMHO that we consider "not seen" an increase

I don't think we consider "not seen" an increase? If we don't see any transitions after a new plank pod is created there is no break in monotonicity, the timeseries doesn't exist for the new plank pod until we do see a transition so the value doesn't decrease until then.

> re 1: Not sure I understand, jobs get created with `.status.state` set to `triggered` already, why would we count they multiple times?

I'm assuming we want to count the creation of a PJ as a state transition to the `triggered` state. Since we are counting from plank it is possible we process the triggered job multiple times before it transitions to `pending`. If we don't keep a cache of the jobs that we processed the last sync loop we won't be able to distinguish newly `triggered` jobs from `triggered` jobs that we have already seen. So that would be problematic for my plan to increment the counter when the state transitions occur rather than retrospectively.


> re 2.: I think this was a design decision. The jenkins controller e.G. also publishes this metric for jenkins jobs. The idea was apparently that the `agent` is inferred from the `job` label.

Interesting, I guess that works just as well as an `agent` label, but it does require us to duplicate the state transition metric logic across all execution agent controllers (today that isn't too bad, but it will be more repetitive with pipelines).

> We already have such a component in prow/cmd/exporter.

Ah I recalled having a metrics component, but I was looking under a different name. I think I'd prefer for the exporter to handle this type of metric generically across execution agents, but I agree that things are mostly working now so the ROI will be better once we have the pipeline controller. 
There are still two inaccuracies in this metric 1) due to the scrape race when the pod starts and 2) in the prow-controller-manager due to missing transitions if multiple happen during the 30 second poll period (namely missing `triggered` transitions), but I think we can live with those for the time being.
